### PR TITLE
reuseable jadbDevices and the shell command returns a string

### PR DESCRIPTION
**Pull request by tg44**

### CODE REVIEW

**Comment by tg44** [comment]
shell execution now can return as byte array too, now the screencap -p saving a correct png:
http://blog.shvetsov.com/2013/02/grab-android-screenshot-to-computer-via.html

**Comment by vidstige** [comment]
Hi,
Thank you for the pull request. :-) What problems are you having with screenshots? Normally you create screenshot functionality on top of jadb, something like [this](https://github.com/vidstige/UiMutilator/blob/master/src/se/vidstige/android/uimutilator/UiDevice.java#L53.

Can you describe more in detail what problems you are facing taking and transferring screenshot?

* `JadbDevice` is not meant to be reusable. If you want to do multiple request you should keep the connection open, if you want to reconnect you should instead create a new `JadbDevice`. Perhaps the class should be called `JadbDeviceConnection`.
* What is the point of the "common" dependency?

**Comment by tg44** [comment]
For the screenshots:
Your implementation is the hard and costly way. You are using 3 adb command to get 1 picture + its not threadsafe... The screenshot making in my link above is run about 2s, your implementation on my device is run about 4-6s due to a filewrites filereads and converts.

The JadbDevice:
So as you mention your jadbDevice is more like a device-connection. I bring the connection inside to the device, so I can recreate the transport object every time when I use a command. As I see every time I use a push or pull twice in the same jadbdevice i get errors, so I think the Transport object and the adb protocol allow only one device call per connection.
Every time I droped/recreated my jadbcDevice nobody closed my 2 streams in Transport. So I do this now too, with the reOpenTransport, witch is called whenever I execute a pull/put/shell/list command.
With this implementation its more like a device for me, I create once, than I can execute more shellcommands on it in sequence, and I can finally read back the answear.

**Comment by vidstige** [comment]
I see, jadb does not seem to handle the fact that sync requests closes the socket. I think for these (list, push, pull, shell iirc) jadb should first open a fresh new transport. Let's do it so that the JadbDevice takes TransportFactory interface as parameter with a single `createTransport()` function, and then let the `JadbConnection` implement this. How about that? 

Also what is the point of the `repairTransportedArray` thingy?

**Comment by tg44** [comment]
Yapp what I did is something similar to the thing you write, I just didn't clean it as many as possible (I hate using interfaces and traits just to one thing, or just to cover up something...).
My point was not "start a new connection before I do my job", more likely "after I do my job I need to try to recreate the things I used". Thats why I use reopenTransport instead of createTransport, I didn't want to restructure your whole code, I just needed the functionality :D And as you can see, I use the JadbConnection exactly as a TransportFactory.

I can't repeat myself anymore about the repair thing :D Read this pls. Its explains what the adb do with raw byte streams and how you can repair it:
http://blog.shvetsov.com/2013/02/grab-android-screenshot-to-computer-via.html
I'm just implemented that in the java way instead of the grep/perl/sed/awk/"whatever linux string tool" way.

**Comment by tg44** [comment]
Additionally, if you are really want to make this project back to life, a maven/ant/sbt dependency tool, a well deserved place in the mavencentral and some intellij and eclipse specific file cleanup woud be nice. I help you if you want :)

**Comment by adelolmo** [comment]
I've just cloned jadb today just to add maven support. I won't create a pull request though, because I pretty much change every possible thing to suit my taste, buy you can definitely grab whatever you need from here: https://github.com/adelolmo/jadb
I'd love to have the jadb library in maven central too!
Thanks for you work :)

**Comment by vidstige** [comment]
Yeah, I saw what you had done, but let's make it even clearer in the code. I'd rather we restructure a little bit to keep consistent. :-) All the other function that creates some thing are called "create" something, so let's rename this also to keep it consistent, and I would prefer to add the interface to make it clear that the Device class _only_ uses the createTransport. I would also make it so that it is created before a file operation and then automatically closed after by the server, this way one connection can be kept open for additional commands. Also the mocked server should probably be updated to reflect this so this behaviour can be tested in the unit tests.

I would actually like this to be made more like a general stdout thingy, so that you can execute any command and pipe the output to e.g. a StreamOutput or whatever that is called in java again (mostly coding js these days). The the screenshot thingy can be moved to the RealDeviceTests as an example or so.

If you make those refactorins I'd be glad to merge! :-) Great addition.

**Comment by vidstige** [comment]
And yes, feel free to package and add to maven repo. I would be greatful! Will accept pull requests as needed.

**Comment by vidstige** [comment]
@adelolmo alright, thanks for your feedback. Will look into your fork to learn how people are using this project.

**Comment by vidstige** [comment]
Also for bonus, I think it's more java-like to do the filtering of newlines using a `StreamReader`. This also greatly reduces memory usage and latency when piping large files. Something similar to [this](http://stackoverflow.com/a/14754606/363437).

**Comment by tg44** [comment]
I cleand up a bit, still need a bit cleaner Device and Transport.

@adelolmo if you make a pullreq only with the filecleanups and maven, I think that woud be mergable anytime besides my modifications

File: .idea/misc.xml [hunk 1]
```diff
@@ -3,8 +3,8 @@
   <component name="EntryPointsManager">
     <entry_points version="2.0" />
   </component>
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_6" assert-keyword="true" jdk-15="true" project-jdk-name="1.6" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_6" default="false" assert-keyword="true" jdk-15="true" project-jdk-name="1.7" project-jdk-type="JavaSDK">
     <output url="file://$PROJECT_DIR$/out" />
   </component>
-</project>
-
+  <component name="WebServicesPlugin" addRequiredLibraries="true" />
```
Comment (vidstige) [inline .idea/misc.xml]: what does this do?

File: .idea/misc.xml [hunk 2]
```diff
@@ -3,8 +3,8 @@
   <component name="EntryPointsManager">
     <entry_points version="2.0" />
   </component>
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_6" assert-keyword="true" jdk-15="true" project-jdk-name="1.6" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_6" default="false" assert-keyword="true" jdk-15="true" project-jdk-name="1.7" project-jdk-type="JavaSDK">
```
Comment (vidstige) [inline .idea/misc.xml]: what does this do? Add JDK 1.7 support?

File: src/se/vidstige/jadb/JadbConnection.java [hunk 3]
```diff
@@ -5,15 +5,13 @@
 import java.util.ArrayList;
 import java.util.List;
 
-public class JadbConnection {
+public class JadbConnection implements ITransportFactory{
```
Comment (vidstige) [inline src/se/vidstige/jadb/JadbConnection.java]: whitespace before brace

File: src/se/vidstige/jadb/JadbDevice.java [hunk 4]
```diff
@@ -6,31 +6,27 @@
 
 public class JadbDevice {
 	private final String serial;
-	private final Transport transport;
-	private boolean selected = false;
+	private Transport transport;
+    private final ITransportFactory tFactory;
```
Comment (vidstige) [inline src/se/vidstige/jadb/JadbDevice.java]: weird name? Let's spell it out like so `transportFactory`

File: src/se/vidstige/jadb/JadbDevice.java [hunk 5]
```diff
@@ -6,31 +6,27 @@
...
+        this.tFactory = tFactory;
     }
 
-    private void ensureTransportIsSelected() throws IOException, JadbException {
-        if (!selected)
-        {
-            selectTransport();
-            selected = true;
+    private void getTransport() throws IOException, JadbException {
+        if(transport!=null && !transport.isClosed()){
```
Comment (vidstige) [inline src/se/vidstige/jadb/JadbDevice.java]: whitespace before parentesis and before brace as well as around not-equals, like so 

if (transport != null && !transport.isClosed()) {

File: src/se/vidstige/jadb/JadbDevice.java [hunk 6]
```diff
@@ -6,31 +6,27 @@
...
-        this.transport = transport;
+        this.tFactory = tFactory;
     }
 
-    private void ensureTransportIsSelected() throws IOException, JadbException {
-        if (!selected)
-        {
-            selectTransport();
-            selected = true;
+    private void getTransport() throws IOException, JadbException {
```
Comment (vidstige) [inline src/se/vidstige/jadb/JadbDevice.java]: normally in java methods named "get" returns something. Here the method ensures a valid exists, so let's rename to something more appropriate.

File: src/se/vidstige/jadb/JadbDevice.java [hunk 7]
```diff
@@ -50,28 +46,39 @@ public String getSerial()
...
+            // TODO: quote arg if it contains space
+            shellLine.append(arg);
+        }
+        send("shell:" + shellLine.toString());
+    }
+
     public List<RemoteFile> list(String remotePath) throws IOException, JadbException {
-        ensureTransportIsSelected();
+        getTransport();
         SyncTransport sync  = transport.startSync();
```
Comment (vidstige) [inline src/se/vidstige/jadb/JadbDevice.java]: one whitespace too much before =

File: src/se/vidstige/jadb/Transport.java [hunk 8]
```diff
@@ -8,6 +11,11 @@
 
 	private final OutputStream outputStream;
 	private final InputStream inputStream;
+	private boolean closed=false;
```
Comment (vidstige) [inline src/se/vidstige/jadb/Transport.java]: whitespace around assign

File: src/se/vidstige/jadb/Transport.java [hunk 9]
```diff
@@ -8,6 +11,11 @@
...
+
+	public boolean isClosed(){
```
Comment (vidstige) [inline src/se/vidstige/jadb/Transport.java]: whitespace

**Comment by vidstige** [comment]
Alright, great! Thanks for your contribution. There are some minor whitespace consistence issues, etc. Also the bytearray thingy can be cleaned up. It's way better than current master though, so I'm ready to merge soon. Will fix up whitespace etc this week if I get some time over.

**Pull request merged**